### PR TITLE
fix issue1: update for current core_kernel and lwt

### PR DESCRIPTION
**Pull request by dbrumley**
Fixes labeled arguments for `bind` compilation error in issue 1 for modern core_kernel and lwt.

### CODE REVIEW

**Comment by dbrumley** [comment]
that's weird. the travis build problem is the one I thought the patch fixed...at least on my system.

File: lib/core_lwt.ml [hunk 1]
```diff
@@ -9,7 +9,7 @@ module Std = struct
     include Core_lwt_extra
     module Main = Lwt_main
     module Unix = Lwt_unix
-    module Chan = Lwt_chan
+    module Chan = Lwt_io
```
Comment (ivg) [inline lib/core_lwt.ml]: This substitution is not valid, `Chan` is a module alias for `Lwt_chan`, and the same is true for other modules. We just moving here from the pseudo `Lwt_` namespace, to a modular `Lwt.`, e.g., `Lwt_xxx` is now `Lwt.Xxx`.

**Review by ivg — CHANGES_REQUESTED** [review]
You need to update the opam file and constrain the lower bound of the core_kernel version. It should be something above v0.9. 

The build fails because it grabs core.113.00 that doesn't use a labeled parameter for the `bind` function.

File: opam [hunk 2]
```diff
@@ -25,7 +25,7 @@ depends: [
     "oasis" {build & >= "0.4.0"}
     "base-unix"
     "base-threads"
-    "core_kernel"
+    "core_kernel" { >= "0.9.0"}
```
Comment (ivg) [inline opam]: it should be `v0.9`, notice the leading `v`.

File: lib/core_lwt.ml (hunk 1 above)
Comment (dbrumley) [reply lib/core_lwt.ml]: During build without the change:
```
File "lib/core_lwt.ml", line 12, characters 18-26:
Warning 3: deprecated: module Lwt_chan
 Use module Lwt_io.
```
Do you want to rename `Chan` to `Io` for the substitution?

File: opam (hunk 2 above)
Comment (dbrumley) [reply opam]: i was wondering that and was going to let travis tell me, but this is quicker. (It did build on my system).

Comment (dbrumley) [reply opam]: I'll change that after the above is answered re: `Chan`.

File: lib/core_lwt.ml (hunk 1 above)
Comment (ivg) [reply lib/core_lwt.ml]: We can either drop Chan at all, or keep it deprecated. I think that we can easily just remove it.

**Comment by ivg** [comment]
OK, now it fails because the OCAML compiler version is too old. Although it is set to the "latest", apparently the latest value is hardcoded somewhere in the scripts upstream, so it is no longer the latest. My suggestion is to set it explicitly to 4.05.0 in travis.yml (see the environment variable OCAML_VERSION in the env stanza).

**Comment by dbrumley** [comment]
Set to 4.03.0 as that's what is working on my system. Fingers crossed.

**Comment by dbrumley** [comment]
Alright, I'm confused here. Travis is failing with:
```
  - core_kernel>=v0.9.0 is not available because your system doesn't comply with ocaml-version >= "4.04.1".
```

Yet on my system:
```
(py36) Davids-MBP:core-lwt dbrumley$ opam info core_kernel  | head -2
             package: core_kernel
             version: v0.9.0
(py36) Davids-MBP:core-lwt dbrumley$ ocamlc --version
4.03.0
```

I could bump to 4.05.0 as originally suggested, but would love to know why it seems to work fine with 4.03.0 on my system with v0.9.0 core_kernel.

**Comment by ivg** [comment]
I think that it is just a failure in OPAM, because aspcud is not installed and in its absence a heuristic is used ... that fails more often than succeeds. The ASP problem, that OPAM is trying to solve, is undecidable in general, and given that specifications in the opam-repository is an approximation, that is both over-constrained (for some packages) and under-constrained (for other packages) the solution is often far from perfect. 

The solution that we see here is really funny, I've greped the opam-repository for any occurrences of  4.04.1 and the only package that states such constraint is some `horned_worm` package, that is not used by any other packages at all. So where this constraint did come from, I have no idea. Probably, it is just some widening.

So the main problem is that the `aspcud` package is not installed, and the proper solution would be to add it into the `packages` section in the `.travis.yml`

**Comment by ivg** [comment]
damn, now I'm frustrated)) Of course, I can't reproduce it on my system. Though I have exactly the same setup as on Travis.

**Comment by dbrumley** [comment]
Is it possible opam package information is cached? Couldn't tell from build if it was a new switch or installed and updated from scratch. ??????

**Comment by ivg** [comment]
> Is it possible opam package information is cached?

No, it's not, caching is disabled for this repository. This was the first thing I've checked. 

So my last commit fixed this issue, though we have the new one:
```
Error: The implementation lib/core_lwt_pool.ml
# [...]
3m#            ?validate:('a -> bool Lwt.t) ->
#            ?check:('a -> (bool -> unit) -> unit) ->
#            ?dispose:('a -> unit Lwt.t) -> (unit -> 'a Lwt.t) -> 'a t
#        is not included in
#          val create :
#            int ->
#            ?check:('a -> (bool -> unit) -> unit) ->
#            ?validate:('a -> bool Lwt.t) -> (unit -> 'a Lwt.t) -> 'a t
```

Apparently, it is because you have a different version of Lwt on your system, and in version that is used in Travis there is a new parameter called `dispose` and no longer have an unnamed parameter of type `int` (whatever it is). 

The solution would be to add explicit constraint to the lwt version. You can use the `=` constraint it put the version that you're using, alternatively, you can use a range, e.g., `{ >= "2.0.0" & < "3.0.0"}`. Anton respects semantic versioning, so if your code compiles with the 2.0 then it should compile with  all 2.x. (I'm not sure are you using 1.x, or 2.x, or 3.x so make sure that choose the proper version)

**Comment by dbrumley** [comment]
Alright, seems the original issue is with bap using v0.9.0 of core_kernel, which is used by the current development version of bap.  

The current version of core-lwt works fine with bap-1.3 as in opam, as that depends on core_kernel 1113.33.xx.  (There was an error in my build where it was using an outdated `setup.data` with a bad dependency.)

I'm going to test a build just to verify everything works with bap 1.3 and core_kernel v0.9.0. It's unclear whether this patch should be pushed now or wait even if it builds.

**Review by ivg — APPROVED** [review]
👍

**Pull request merged**